server/grpc/reflection/v2alpha1/reflection.go

```diff
@@ -0,0 +1,219 @@
+			impls[j] = impl.TypeUrl
+		}
+		ifaceImplementers[iface.Fullname] = impls
+	}
+	return reflectionServiceServer{
+		desc: desc,
+	}, nil
+}
+
+// newCodecDescriptor describes the codec given the codectypes.InterfaceRegistry
+func newCodecDescriptor(ir codectypes.InterfaceRegistry) (*CodecDescriptor, error) {
+	registeredInterfaces := ir.ListAllInterfaces()
+	interfaceDescriptors := make([]*InterfaceDescriptor, len(registeredInterfaces))
+
+	for i, iface := range registeredInterfaces {
+		implementers := ir.ListImplementations(iface)
+		interfaceImplementers := make([]*InterfaceImplementerDescriptor, len(implementers))
+		for j, implementer := range implementers {
+			pb, err := ir.Resolve(implementer)
+			if err != nil {
+				return nil, fmt.Errorf("unable to resolve implementing type %s for interface %s", implementer, iface)
+			}
+			pbName := proto.MessageName(pb)
+			if pbName == "" {
+				return nil, fmt.Errorf("unable to get proto name for implementing type %s for interface %s", implementer, iface)
+			}
+			interfaceImplementers[j] = &InterfaceImplementerDescriptor{
+				Fullname: pbName,
+				TypeUrl:  implementer,
+			}
+		}
+		interfaceDescriptors[i] = &InterfaceDescriptor{
+			Fullname: iface,
+			// NOTE(fdymylja): this could be filled, but it won't be filled as of now
+			// doing this would require us to fully rebuild in a (dependency) transitive way the proto
+			// registry of the supported proto.Messages for the application, this could be easily
+			// done if we weren't relying on gogoproto which does not allow us to iterate over the
+			// registry. Achieving this right now would mean to start slowly building descriptors
+			// getting their files dependencies, building those dependencies then rebuilding the
+			// descriptor builder. It's too much work as of now.
+			InterfaceAcceptingMessages: nil,
+			InterfaceImplementers:      interfaceImplementers,
+		}
+	}
+
+	return &CodecDescriptor{
+		Interfaces: interfaceDescriptors,
+	}, nil
+}
+
+func newQueryServiceDescriptor(srv *grpc.Server) *QueryServicesDescriptor {
+	svcInfo := srv.GetServiceInfo()
+	queryServices := make([]*QueryServiceDescriptor, 0, len(svcInfo))
+	for name, info := range svcInfo {
+		methods := make([]*QueryMethodDescriptor, len(info.Methods))
+		for i, svcMethod := range info.Methods {
+			methods[i] = &QueryMethodDescriptor{
+				Name:          svcMethod.Name,
+				FullQueryPath: fmt.Sprintf("/%s/%s", name, svcMethod.Name),
+			}
+		}
+		queryServices = append(queryServices, &QueryServiceDescriptor{
+			Fullname: name,
+			Methods:  methods,
+		})
+	}
+	return &QueryServicesDescriptor{QueryServices: queryServices}
+}
+
+func newTxDescriptor(ir codectypes.InterfaceRegistry) (*TxDescriptor, error) {
+	// get base tx type name
+	txPbName := proto.MessageName(&tx.Tx{})
+	if txPbName == "" {
+		return nil, fmt.Errorf("unable to get *tx.Tx protobuf name")
+	}
+	// get msgs
+	msgImplementers := ir.ListImplementations(sdk.MsgInterfaceProtoName)
```

Comment:
One idea I just thought (for the future) is creating new Protobuf options used for reflection, e.g.

```proto
// cosmos.bank.v1beta1
service Msg {
  option cosmos_msg_service = true;
  // all methods
}

service Query {
  option cosmos_query_service = true;
  // all methods
}
```

Which would avoid using `ListImplementations` for getting all ServiceMsgs. Just an idea.

Reply:
Yes, having this would sure help. Although I feel the best would be that the configurator becomes part of *BaseApp and we explicitly mark query services and msg services for the application.